### PR TITLE
ER-380 Prevent users navigating to invalid steps in the user journey

### PR DESCRIPTION
**Pull request by chrisjensenuk**
To test locally turn off the developer exception page and use the error handler
//app.UseDeveloperExceptionPage();
app.UseExceptionHandler("/error/handle");

### CODE REVIEW

**Comment by chrisjensenuk** [comment]
I'm not overly happy about it. The utility class was introduced in a previous story so I just extended the pattern.  I could add an Employer.Web specific extension to VacancyClient or introduce a wrapper class around VacancyClient.  What are people's thoughts/preferences/ideas?

File: src/Employer/Employer.Web/Utility.cs [hunk 1]
```diff
@@ -2,15 +2,90 @@
...
+                redirectRoute, new { vacancy.EmployerAccountId, VacancyId = vacancy.Id });
+        }
+
+        private static IList<string> GetValidRoutesForVacancy(Vacancy vacancy)
+        {
+            var validRoutes = new List<string>();
+
+            validRoutes.AddRange(new [] {RouteNames.Title_Post, RouteNames.Title_Get});
+            if (string.IsNullOrWhiteSpace(vacancy.Title))
+                return validRoutes;
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Utility.cs]: We need to return a collection as there are many valid routes.
1) User creates new vacancy
2) User progresses to ShortDescription
3) User clicks back in the browser to make a change to the Employer address
4) They should land on the employer screen and not be redirected back to ShortDescription

File: src/Shared/Recruit.Vacancies.Client/Domain/Exceptions/InvalidRouteForVacancyException.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Esfa.Recruit.Vacancies.Client.Domain.Exceptions
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Shared/Recruit.Vacancies.Client/Domain/Exceptions/InvalidRouteForVacancyException.cs]: 👍

File: src/Employer/Employer.Web/Utility.cs [hunk 3]
```diff
@@ -2,15 +2,90 @@
...
+            return vacancy;
+        }
+
+        public static async Task<Vacancy> GetAuthorisedVacancyAsync(IEmployerVacancyClient client, Guid vacancyId, string employerAccountId, string routeName)
+        {
+            var vacancy = await client.GetVacancyAsync(vacancyId);
+
+            CheckAuthorisedAccess(vacancy, employerAccountId);
+
+            CheckRouteIsValidForVacancy(vacancy, routeName);
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Utility.cs]: As we post to a route I think we should validate post routes too.

File: src/Employer/Employer.Web/Utility.cs [hunk 4]
```diff
@@ -2,15 +2,90 @@
...
+            validRoutes.AddRange(new[] { RouteNames.ShortDescription_Post, RouteNames.ShortDescription_Get});
+            if (string.IsNullOrWhiteSpace(vacancy.ShortDescription))
+                return validRoutes;
+
+            validRoutes.AddRange(new[] { RouteNames.Training_Post, RouteNames.Training_Get});
+            if (string.IsNullOrWhiteSpace(vacancy.ProgrammeId))
+                return validRoutes;
+
+            validRoutes.AddRange(new[] { RouteNames.Wage_Post, RouteNames.Wage_Get});
+            if (vacancy.Wage?.WageType == null)
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Utility.cs]: There is some cross-over here. Maybe CheckHasCompletedLastStepOfPart1 should be using this route validation instead. I'll have a look.

File: src/Employer/UnitTests/Employer.Web/UtilityTests.cs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+        private void CheckRouteIsValidForVacancyTest(Vacancy vacancy, string route, bool shouldRedirect, string expectedRedirectRoute)
+        {
+            bool isRedirecting = false;
+            string actualRedirectRoute = null;
+            string actualEmployerAccountId = null;
+            Guid actualVacancyId;
+
+            try
+            {
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Employer/UnitTests/Employer.Web/UtilityTests.cs]: 👍

File: src/Employer/Employer.Web/Utility.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sachpatel) [reply src/Employer/Employer.Web/Utility.cs]: what I suggested would have put us in a infinite loop 😝

File: src/Employer/Employer.Web/Controllers/ErrorController.cs [hunk 6]
```diff
@@ -49,7 +49,6 @@ public IActionResult ErrorHandler()
 
             if (exceptionFeature != null)
             {
-                string routeWhereExceptionOccurred = exceptionFeature.Path;
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Controllers/ErrorController.cs]: Will fix

File: src/Employer/Employer.Web/Utility.cs [hunk 7]
```diff
@@ -2,15 +2,90 @@
...
+
+            if (validRoutes == null || validRoutes.Contains(currentRouteName))
+            {
+                return;
+            }
+
+            var redirectRoute = validRoutes.Last();
+            
+            throw new InvalidRouteForVacancyException(string.Format(ExceptionMessages.RouteNotValidForVacancy, currentRouteName, redirectRoute),
+                redirectRoute, new { vacancy.EmployerAccountId, VacancyId = vacancy.Id });
```
[the remark this answers is not in the document]
Comment (chrisjensenuk) [reply src/Employer/Employer.Web/Utility.cs]: I was hoping not to couple the exception to the VRM however this added complexity is probably not worth it as this exception is always providing the EmplyerAccountId and VacancyId in any case.  Will change to use VacancyRouteModel. 👍

**Pull request merged**